### PR TITLE
build: use semantic-release to automate release

### LINKED ISSUE #1647: Add pull request template and policy
https://help.github.com/articles/about-issue-and-pull-request-templates/

Possible template/policies:
- PR title, which becomes the message of squashed merge commit
- Linking the issue with correct syntax (closing word)
- Adding an entry to CHANGELOG.md
- Updating `docs/supported.md`, `docs/support_table.md`, and other documentation if applicable
- Adding unit/regression/screenshot tests

### LINKED ISSUE #1780: Versioned docs are rendered using latest KaTeX
Currently, versioned docs are rendered using latest built KaTeX.

~This will be solved by Docusaurus 2, as it snapshots versioned docs.~

### LINKED ISSUE #2011: Use semantic-release to simplify releases
semantic-release determines what version to bump to when releasing based on commit messages, see https://github.com/semantic-release/semantic-release for details.  This may take a bit of work to integrate this into our existing release script.

As a first step though we could adopt https://github.com/semantic-release/commit-analyzer which would enforce a specific commit message style which would be helpful in manually determining which version to bump to.

### PR DESCRIPTION
**Pull request by ylemkimon**
Closes #2011. I'm really excited and surprised that it's really simple to use the `semantic-release`.

## Previous release process
* Run `release.sh` with the version to release
  * (1) Create a release branch
  * (2) Update `package.json` to the version to release
  * (3) Build
  * (4) Update `package.json` to the next version
  * (5) Update the README and documentation
  * (6) Generate a new version of the docs
  * (7) Make a release commit and push
* (8) Update CHANGELOG
* (9) Open PR
* (10) Update SRI hashes if files are changed
* Run `release.sh -p`
  * (11) Update `package.json` to the version to release
  * (12) Build distribution (`yarn dist`)
  * ~(13) Make another release commit~ **Removed in #2385**
  * (14) Make a tag and push
  * (15) Publish to `npm`
* (16) Create a GitHub release and upload artifacts

## New release process (fully automated)
* A PR is merged to a release branch, e.g., `master` ([Triggering a release](https://github.com/semantic-release/semantic-release#triggering-a-release))
* Determine the type of release based on commits added since the last release. ([@semantic-release/commit-analyzer](https://github.com/semantic-release/commit-analyzer))
  * [x] **All PRs should follow [Angular Commit Message Conventions](https://github.com/angular/angular.js/blob/master/DEVELOPERS.md#-git-commit-guidelines)**
  * [x] **Updated CONTRIBUTING.md and added a pull request template (Fixes #1647)**
  * [x] ~@kevinbarabash I've sent a request to install https://github.com/apps/semantic-pull-requests to enforce the convention~
* Generate release notes for commits added since the last release. ([@semantic-release/release-notes-generator](https://github.com/semantic-release/release-notes-generator))
* (8) Update CHANGELOG ([@semantic-release/changelog](https://github.com/semantic-release/changelog))
* (2, 11) Update `package.json` to the version to release ([@semantic-release/npm](https://github.com/semantic-release/npm))
  * `postversion`
    * (3, 12) Build distribution (`yarn dist`)
    * (5) Update the README and documentation
* (7) Make a release commit and push ([@semantic-release/git](https://github.com/semantic-release/git))
  * [x] **Permission to push to protected branch is required**
  * [ ] ~@kevinbarabash Could you add [GitHub token](https://docs.github.com/en/github/authenticating-to-github/creating-a-personal-access-token) with `public_repo` permission and [npm token](https://docs.npmjs.com/creating-and-viewing-authentication-tokens) to [GitHub Actions secrets](https://docs.github.com/en/actions/configuring-and-managing-workflows/creating-and-storing-encrypted-secrets) as `GH_TOKEN` and `NPM_TOKEN`, respectively?~
  * [x] I've added branch protection rules to release branches.
* (14) Make a tag and push ([@semantic-release/git](https://github.com/semantic-release/git))
* (15) Publish to `npm` ([@semantic-release/npm](https://github.com/semantic-release/npm))
* (16) Create a GitHub release and upload artifacts ([@semantic-release/github](https://github.com/semantic-release/github))
* Notify of new releases or errors

### Removed
* (1) Creating a release branch and (9) opening a PR is no longer required
* There is no risk of SRI hashes changing and `semantic-release` is only run after all tests pass. So (10) updating hashes and running tests in (12) `yarn dist` is no longer required.

### Changed
* (6) Versioned docs are removed. They were broken anyway (#1780). Instead, links to previous Netlify deploys are used, https://5f12204e3f16fe932ffc7a05--katex.netlify.app/versions.html. Closes #1780.

Blocked by a new release.

### CODE REVIEW

File: .releaserc [hunk 1]
```diff
@@ -0,0 +1,19 @@
+{
```
Comment (ylemkimon) [inline .releaserc:1]: This

File: package.json [hunk 2]
```diff
@@ -1,12 +1,12 @@
 {
```
Comment (ylemkimon) [inline package.json:1]: and this file are only relevant changes. Other files are mostly side-effects.

**Comment by ylemkimon** [comment]
I've setup a test repo on https://github.com/ylemkimon/KaTeX. https://github.com/ylemkimon/KaTeX/commit/47d16c987780b3c395341c7953569c0dc34b2203 ([npm](https://www.npmjs.com/package/@ylemkimon/katex/v/0.12.1), [GitHub release](https://github.com/ylemkimon/KaTeX/releases/tag/v0.12.1)) is a manual release after #2323 is merged and https://github.com/ylemkimon/KaTeX/commit/fc33eaa60064b860ac714354f8366e409cac7a0a ([npm](https://www.npmjs.com/package/@ylemkimon/katex/v/0.13.1), [GitHub release](https://github.com/ylemkimon/KaTeX/releases/tag/v0.13.1), [CircleCI log](https://app.circleci.com/pipelines/github/ylemkimon/KaTeX/18/workflows/85862b95-2aec-412a-9565-02ccce5e802c/jobs/158/parallel-runs/0/steps/0-107)) is an automated release after this PR and a dummy commit.

**Comment by kevinbarabash** [comment]
Thanks for taking this on!  I'm super excited about semantic-release.  It'll automate the last remaining manual part of the release process and will help to avoid not bumping things when they should be bumped.  I'll try to do the steps that you flagged me on sometime this week. 🤞

**Comment by kevinbarabash** [comment]
I was looking at some dependabot PRs and that got me wondering: is there a way to config dependabot PR titles to conform to Angular Commit Message Conventions?

**Comment by kevinbarabash** [comment]
I've approved the request for the "Semantic Pull Requests" GitHub App.

**Comment by ylemkimon** [comment]
@kevinbarabash Thank you!
> I was looking at some dependabot PRs and that got me wondering: is there a way to config dependabot PR titles to conform to Angular Commit Message Conventions?

According to docs, GitHub Dependabot attempts to detect your commit message preferences and use similar patterns, by default. I was curious about the logic, so I looked at [the source](https://github.com/dependabot/dependabot-core/blob/92533d45e69aab0aa5b3bccf90b7668ed678b476/common/lib/dependabot/pull_request_creator/pr_name_prefixer.rb#L64-L79), and it turns out that it figures out what convention the repo is using by analyzing a previous Dependabot commit or existing commits.

So if one once merges a Dependabot PR with Angular convention, Dependabot automatically uses the convention afterwards.

**Comment by kevinbarabash** [comment]
@ylemkimon do you still need me to do add a GitHub access token to CircleCI?

**Comment by ylemkimon** [comment]
@kevinbarabash Yes, it needs the token with admin privileges to override the branch protection.

**Comment by kevinbarabash** [comment]
@ylemkimon it looks like we already have a `GH_TOKEN` in the CircleCI project settings but my guess is that it doesn't have the right scopes.  Do you want me to replace that token with the one you requested above or would you like me to add a new one with a different name?  If you want a new one, what should I name it?

**Comment by ylemkimon** [comment]
(Posted with a wrong account) @kevinbarabash I've added my token to test this PR. You can replace it.

**Comment by kevinbarabash** [comment]
I think this PR should be blocked on #2417 since that one will be disabling CircleCI which will have a significant impact on this.

**Comment by ylemkimon** [comment]
@kevinbarabash I've updated the PR to use GitHub Actions.

Regarding tokens, after careful thoughts, I think it'd be best to create a separate account with the access to only this repo, to minimize damage if they are compromised.

**Comment by kevinbarabash** [comment]
@ylemkimon good call.  It's too bad there's no way to create a personal access token scoped to an individual repo.  Creating a separate account just for this is a good workaround.

**Comment by ylemkimon** [comment]
@kevinbarabash I have the username @KaTeX-bot. I'll change the username, if you'd like to use it.

**Comment by kevinbarabash** [comment]
`@KaTeX-bot` sounds good to me.  Thanks for creating a user to help with publishes.  Can you add them as a member of the `KaTeX` org?

**Comment by kevinbarabash** [comment]
@ylemkimon is everything that's need for step 7 satisfied now?  Is this ready for review now?  Is there a way to test this, maybe to build a `-test` release or something?

**Comment by ylemkimon** [comment]
@kevinbarabash I've re-forked the repo on [ylemkimon/KaTeX](https://github.com/ylemkimon/KaTeX). https://github.com/ylemkimon/KaTeX/commit/43495b2040df24a2c2ea6fe5b348cbad541e83d0 ([npm](https://www.npmjs.com/package/@ylemkimon/katex/v/0.14.0), [GitHub release](https://github.com/ylemkimon/KaTeX/releases/tag/v0.14.0)) is a manual release and https://github.com/ylemkimon/KaTeX/commit/98b1f8020003f46ff7d49d75406f946dc7e5ae15 ([npm](https://www.npmjs.com/package/@ylemkimon/katex/v/0.14.1), [GitHub release](https://github.com/ylemkimon/KaTeX/releases/tag/v0.14.1), [GitHub Actions log](https://github.com/ylemkimon/KaTeX/runs/1050206131)) is an automated release after this PR and a dummy feature commit. It'd be also possible to [publish a pre-release version by pushing to its branch](https://github.com/semantic-release/semantic-release/blob/master/docs/recipes/pre-releases.md#publishing-pre-releases) once this PR is merged.

This PR is still blocked by one last manual release. Otherwise, if you're okay with using my accounts, this PR is ready for review.

**Comment by kevinbarabash** [comment]
> This PR is still blocked by one last manual release. Otherwise, if you're okay with using my accounts, this PR is ready for review.

The releases to https://github.com/ylemkimon/KaTeX and https://www.npmjs.com/package/@ylemkimon/katex were just for testing purposes, right?  In terms of using your accounts, was that just for testing the release process or will that be necessary for the release process once this is merged?

**Comment by ylemkimon** [comment]
@kevinbarabash 
> The releases to https://github.com/ylemkimon/KaTeX and https://www.npmjs.com/package/@ylemkimon/katex were just for testing purposes, right?

Yes, we need a real KaTeX release.

> In terms of using your accounts, was that just for testing the release process or will that be necessary for the release process once this is merged?

It'd be necessary for the release process. It's possible to set them in https://github.com/KaTeX/KaTeX/settings/secrets.

**Comment by kevinbarabash** [comment]
> It'd be necessary for the release process. It's possible to set them in https://github.com/KaTeX/KaTeX/settings/secrets.

Would it be possible to use @KaTeX-bot for the release process instead of your personal account?

**Comment by ylemkimon** [comment]
@kevinbarabash I meant @KaTeX-bot by my account, i.e., account under my control.

**Comment by ylemkimon** [comment]
It turns out currently, there is no reliable solution to enable semantic-release on the monorepo. I've opened a discussion in #2517.

**Comment by kevinbarabash** [comment]
> Blocked by a new release.

I guess we should do a new release then.  I'd like to try to get the remaining three `fix` PRs reviewed and landed first.  I should be able to do that over the course of the next week.

**Comment by kevinbarabash** [comment]
I've put up a PR for a 0.13.0 release, see #2824.

**Comment by ylemkimon** [comment]
Closed in favor of #2830.

**Pull request closed without merging**